modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java

```diff
@@ -1286,21 +1279,154 @@ private ArrayIndexNormalizeHelper() {}
         public final int numCaptures;
         public final String encoding;
 
+        private static final String FORMAT = "[SD][tf]symbol.methodName,numCaptures";
+
         public Encoding(boolean isStatic, boolean needsInstance, String symbol, String methodName, int numCaptures) {
             this.isStatic = isStatic;
             this.needsInstance = needsInstance;
-            this.symbol = symbol;
-            this.methodName = methodName;
+            this.symbol = Objects.requireNonNull(symbol);
+            this.methodName = Objects.requireNonNull(methodName);
+            this.numCaptures = numCaptures;
+            this.encoding = encoding();
+            validate();
+        }
+
+        private Encoding(boolean isStatic, boolean needsInstance, String symbol, String methodName, int numCaptures, String encoding) {
+            this.isStatic = isStatic;
+            this.needsInstance = needsInstance;
+            this.symbol = Objects.requireNonNull(symbol);
+            this.methodName = Objects.requireNonNull(methodName);
             this.numCaptures = numCaptures;
-            this.encoding = (isStatic ? "S" : "D") + (needsInstance ? "t" : "f") +
-                            symbol + "." +
-                            methodName + "," +
-                            numCaptures;
+            this.encoding = Objects.requireNonNull(encoding);
+            validate();
+        }
+
+        public static Encoding parse(String encoding) {
+            Objects.requireNonNull(encoding);
+            if (encoding.length() < 6) {
+                throw new IllegalArgumentException("Encoding too short. Minimum 6, given [" + encoding.length() + "]," +
+                    " encoding: [" + encoding + "], format: " + FORMAT + "");
+            }
+
+            Encoding parsed = new Encoding(isStatic(encoding), needsInstance(encoding), symbol(encoding), methodName(encoding),
+                numCaptures(encoding), encoding);
+
+            String generatedEncoding = parsed.encoding();
+            if (parsed.encoding.equals(generatedEncoding) == false) {
+                throw new IllegalStateException("generated encoding [" + generatedEncoding + "] must equal given encoding" +
+                    " [" + encoding + "]");
+            }
+
+            return parsed;
+        }
+
+        private void validate() {
```

Comment:
Do you think it's worth calling this in an assertion so it doesn't run all the time?

Reply:
That's my paranoid instinct. I know this code isn't so sensitive we're worried about one extra allocation. But the extra checks for what's "always" true maybe much? I dunno.

Reply:
What do you think about only having it run during compilation but not during runtime?  ie it would not run for the parsing constructor.

Reply:
That sounds good to me. It'd be, like, a public method you'd call to validate?

Reply:
We'll only validate for the non-parsing constructor and, apart from sanity checks, avoid validation in the parsing constructor.